combinetf2/physicsmodels/physicsmodel.py

```diff
@@ -0,0 +1,153 @@
+
+    # function to compute the transformation of the physics model, has to be differentiable.
+    #    For custom physics models, this function should be overridden.
+    #    observables are the provided histograms inclusive in processes: nbins
+    #    params are the fit parameters
+    def compute_flat(self, params, observables=None):
+        return observables
+
+    # function to compute the transformation of the physics model, has to be differentiable.
+    #    For custom physics models, this function can be overridden.
+    #    observables are the provided histograms per process: nbins x nprocesses
+    #    params are the fit parameters
+    def compute_flat_per_process(self, params, observables=None):
+        return self.compute_flat(params, observables)
+
+    # generic version which should not need to be overridden
+    def make_fun(self, fun_flat, params, inclusive=True):
+        compute = self.compute_flat if inclusive else self.compute_flat_per_process
+
+        def fun():
+            if self.need_observables:
+                exp = compute(params, fun_flat())
+            else:
+                exp = compute(params)
+            return exp
+
+        return fun
+
+    # generic version which should not need to be overridden
+    @tf.function
+    def get_data(self, data, data_cov_inv=None):
+        with tf.GradientTape() as t:
+            t.watch(data)
+            output = self.compute_flat(None, data)
+
+        jacobian = t.jacobian(output, data)
+
+        # Ensure the Jacobian has at least 2 dimensions (expand in case output is a scalar)
+        if len(jacobian.shape) == 1:
+            jacobian = tf.expand_dims(jacobian, axis=0)
+
+        if data_cov_inv is None:
+            # Assume poisson uncertainties on data
+            cov_output = (jacobian * data) @ tf.transpose(jacobian)
+        else:
+            # General case with full covariance matrix
+            data_cov = tf.linalg.inv(data_cov_inv)
+            cov_output = jacobian @ data_cov @ tf.transpose(jacobian)
+
+        variances_output = tf.linalg.diag_part(cov_output)
+
+        return output, variances_output, cov_output
+
+
+class PhysicsModelChannel(PhysicsModel):
+    """
+    Abstract physics model to process a specific channel
+    """
+
+    def __init__(self, indata, key, channel):
+        super().__init__(indata, key)
+        channel_info = indata.channel_info[channel]
+        self.channel_info = {channel: channel_info}
+
+        self.start = channel_info["start"]
+        self.stop = channel_info["stop"]
+        self.channel_shape = [len(a) for a in channel_info["axes"]]
+
+    def compute(self, params, observables):
+        return observables
+
+    def compute_per_process(self, params, observables):
+        return self.compute(params, observables)
+
+    def compute_flat(self, params, observables):
+        exp = tf.reshape(observables[self.start : self.stop], tuple(self.channel_shape))
+        return self.compute(params, exp)
```

Comment:
This function now returns the data carrying the shape of the channel, not necessarily flat.  Is this intended or should it be flattened back out here to behave the same as for other models?

Reply:
As long as it's only one channel it's no problem if it's not flat

Reply:
But what happens downstream then?  Does it get flattened out again somewhere?

Reply:
No it's not flattened but if it is flat it gets reshaped into the histogram shape:
https://github.com/WMass/combinetf2/blob/main/combinetf2/workspace.py#L147-L149

Reply:
I think it's better to flatten it to avoid a heterogeneous situation downstream.

This code is already getting a bit complicated for my taste
https://github.com/WMass/combinetf2/blob/c70b43730d788ffff299f8a3b9f5981c82c3ee82/combinetf2/workspace.py#L149-L154

Reply:
OK